sopel/plugins/rules.py

```diff
@@ -0,0 +1,1189 @@
+                del registry[plugin_name]
+                unregistered_rules = unregistered_rules + rules_count
+
+        LOGGER.debug(
+            '[%s] Successfully unregistered %d rules',
+            plugin_name,
+            unregistered_rules)
+
+        return unregistered_rules
+
+    def register(self, rule):
+        """Register a plugin rule.
+
+        :param rule: Rule to register
+        :type rule: :class:`Rule`
+        """
+        with self._register_lock:
+            self._rules[rule.get_plugin_name()].append(rule)
+        LOGGER.debug('Rule registered: %s', str(rule))
+
+    def register_command(self, command):
+        """Register a plugin command.
+
+        :param command: Command to register
+        :type rule: :class:`Command`
+        """
+        with self._register_lock:
+            plugin = command.get_plugin_name()
+            self._commands[plugin][command.name] = command
+        LOGGER.debug('Command registered: %s', str(command))
+
+    def register_nick_command(self, command):
+        """Register a plugin nick command.
+
+        :param command: Nick command to register
+        :type rule: :class:`NickCommand`
+        """
+        with self._register_lock:
+            plugin = command.get_plugin_name()
+            self._nick_commands[plugin][command.name] = command
+        LOGGER.debug('Nick Command registered: %s', str(command))
+
+    def register_action_command(self, command):
+        """Register a plugin action command.
+
+        :param command: Action command to register
+        :type rule: :class:`ActionCommand`
+        """
+        with self._register_lock:
+            plugin = command.get_plugin_name()
+            self._action_commands[plugin][command.name] = command
+        LOGGER.debug('Action Command registered: %s', str(command))
+
+    def has_rule(self, label, plugin=None):
+        """Tell if the manager knows a rule with this ``label``.
+
+        :param str label: Label of the rule to look for
+        :param str plugin: Optional filter on the plugin name
+        :return: ``True`` if the rule exists, ``False`` otherwise
+
+        The optional parameter ``plugin`` can be provided to limit the rules
+        to the ones of said plugin.
+        """
+        rules = (
+            itertools.chain(*self._rules.values())
+            if plugin is None
+            else self._rules.get(plugin, [])
+        )
+        return any(rule.get_rule_label() == label for rule in rules)
+
+    def has_command(self, name, follow_alias=False, plugin=None):
+        """Tell if the manager knows a command with this ``name``.
+
+        :param str label: Label of the rule to look for
+        :param bool follow_alias: Optional filter to follow alias or not
+        :param str plugin: Optional filter on the plugin name
+        :return: ``True`` if the command exists, ``False`` otherwise
+
+        By default, this method doesn't search commands by their aliases. If
```

Comment:
* it feels better/right to me that you have to explicitly ask for the aliases; I honestly don't have a better explanation at this point
* `has_*` methods are for plugin authors and future usages

I'm still trying to find a good interface for the rules manager. I think it's good to have these methods, even though I don't have a direct need for them. I put them here, because I don't want the outside world to tinker with the internal state of the rules manager (not `manager.commands` public attribute for example), but I still want to give some control/access to its state.

Reply:
Let's look at it from the perspective of a possible use-case, then.

Let's say I'm a plugin author who wants, for some reason, to programmatically register a new command. I want to be responsible, so I don't want to register my command if its name is already bound to something. In this case, I'd want aliases to be included, so I don't accidentally register something that will also trigger another handler unintentionally.

I'm having trouble thinking of a case when I (as a plugin author) wouldn't want to know about aliases, honestly. I get that explicitly requesting it "feels better", but the default behavior should reflect what most users of the method will expect. We have no hard data, obviously, but I anticipate most users would want aliases included based on my understanding of what this method (and its friends) is for.

Reply:
As said on IRC, I'll switch to `follow_alias=True` and change tests accordingly.